Src/VsVimShared/Implementation/Settings/IVsStoreSpecializedAccessorsMap.cs

```diff
@@ -0,0 +1,22 @@
+﻿using Microsoft.VisualStudio.Settings;
+
+namespace Vim.VisualStudio.Implementation.Settings
+{
+    /// <summary>
+    /// Interface to abstract "Type specialization" code
+    /// Since in .net the only way to do it now is to use static generics
+    /// Which is not test-friendly if we embed it directly into the class
+    /// This allows to have cleaner code without repeating methods for each type
+    /// While also providing fast access as it doesn't involve reflection
+    /// </summary>
+    public interface IVsStoreSpecializedAccessorsMap
+    {
+        bool Initialized<T>();
+
+        void Map<T>(VsSettingsStoreAccessorsProxy<T> accessors);
+
+        VsSettingGetter<T> GetGetter<T>(WritableSettingsStore store);
+
+        VsSettingSetter<T> GetSetter<T>(WritableSettingsStore store);
+    }

```

Comment:
Why did you take this approach, which seems to have a couple of layers, vs. just wrapping `WritableSettingsStore` in an interface? The design seems to be to allow for betting testing around `WritableSettingsStore` (which is indeed hard to test). But it would seem having a single interface on top of it an a small adapter class would accomplish much the same end here. Is there a reason you preferred this approach?

Reply:
HI @jaredpar . I thought about 2 approaches - using runtime mapping with typeof(), GetType() and Type -> Accessor map and the 2nd one - using this emulated dictionary specialization through static type. This approach is intrinsicly typesafe and faster as it is resolved at jit-compilation time. But this approach requires static generic class. I usually don't like this kind of dependencies as they are harder to test and add complexity so I try to encapsulate it.
The idea is to hide this static class as a private class under a minimalistic facade. This is what VsStoreSpecializedAccessorsMap and StaticMapSpecializationProvider do. They are tricky to test but they are very simple.
The main storage implementations (Inheritors of ISettingsStorage) just take them as a dependency, adding interface here decouples this dependency so the classes are simple and testable.
Now there is an additional part for VS storage, since it has methods like GetBool/String - we need a map between these accessor methods and types (through our specializationMap).
This logic is subject to change as application settings change (when we add new setting types), so I didn't want to put it in VsSettingsStore, which will reduce the need to change this class unless the actual underlying WritableSettingsStore changes in future versions of VS.

This way our main extension point when we add new types is VsStoreStaticSpecializationProvider which doesn't contain any logic beyond accessors mapping.

Single layer of abstraction wouldn't allow to decouple logic from hard static class dependency.

Reply:
I think I'm missing something here. Why do you feel a `static` class was an issue here? There wasn't one before and I'm not sure how adding a simple caching layer necessitated one being added.

Reply:
The interface of ISettingsStore now is Get<T> and Set<T>, but underlying WritableSettingsStore from VS SDK has GetBoolean/GetString/GetInt/GetUint/etc.

Generic interface is nicer in terms of code readability so I went with it. To convert generic interface to this set of type-named methods we could either go with chain of if(typeof(T) == typeof(...)) for each type, which doesn't look very nice, a map from Type to delegate or we could leverage the fact that static generic classes have different instances of fields for different generic types and it's resolved at compile-time (so no getting/comparing types at runtime).

The latter requires static class, and I wanted to have a very lightweight wrapper for that static class so the logic related to static class is minimal and the rest could be easily tested/modified. So the cache doesn't depend on static class at compile time.

If we have only 1 layer of indirection (only ISettingsStore) - we'd either have to use typeof(T) -> Func<string, T> map and compare types in runtime or our ISettingsStorage inheritor would incapsulate both settings-related logic and a static class (if we don't want to have runtime type-checks).

Actually if you don't mind using Castle.Core - after adding relative line numbers support - I could simplify all the code to dynamically generated proxy.

Reply:
Sorry been a busy week. Just getting back to this. 

> Generic interface is nicer in terms of code readability so I went with it

In some cases yes but I'm not convinced in this case. The problem is the underlying store here is not generic. It's fixed to a constrained set of types: `int`, `bool`, `string` and `uint`. Wrapping this with a generic interface implies that the storage can handle any type. That means either you must do one of the following:

1. Throw when a type is specified that isn't represented in the underlying storage. 
1. Implement a serialization layer by which any non-natively supported type is serialized, and deserialized, to one of the supported ones.

This PR takes both approaches here. This means while the interface reads as generic and flexible the implementation is actually quite constrained. There are only a small subset of types available that work here. That's why I think the non-generic one is a bit cleaner here. It may have a few more methods but support declaration is clear.

Reply:
I gave it a second thought - What do you think about using some config library (like Config.net)?
This way we'll only need to implement a class that reads/writes strings from/to the default VsStore, Wrapper for CommandKeyBinding serializer and a small wrapper for notifications.
Overall amount of code should be much less and we get serialization support for a lot of types for free.

Reply:
What types do we need to serialize though? Most all of the settings are just the types natively supported and CommandKeyBinding should already round trip to `string` pretty easily.

Reply:
Serialization is just a benefit, main idea here was to reuse existing tools for settings storage/caching. 
I could prepare a separate PR with config.net as a settings layer (it's a small change, and it'd be easier to compare it when it's visual) and you could decide whether you prefer that one or want to continue with modifying the current one.
If you prefer to continue with the current one - I can change interface to Get/Set Bool/Int/String/etc + generic Get/Set that have IFormatter as an extra argument.

Reply:
Lets try that that approach of separating it out.

Reply:
I did some changes so vs settings store is one class now.